btk/create_blend_generator.py

```diff
@@ -1,94 +1,43 @@
-    blend_catalog["ra"] += dx
-    blend_catalog["dec"] += dy
-    return blend_catalog
-
-
-def generate(Args, catalog, sampling_function=None):
-    """Generates a list of blend catalogs of length Args.batch_size. Each blend
-    catalog has entries numbered between 1 and Args.max_number, corresponding
-    to overlapping objects in the blend.
-
-    Args:
-        Args: Class containing input parameters.
-        catalog:
-        sampling_function: Function to sample input catalog from which to draw
-                           blends.
-
-    Yields:
-        Generator for parameters of each galaxy in blend.
-    """
-    while True:
+class BlendGenerator:
+    def __init__(self, catalog, sampling_function, batch_size=8, verbose=False):
+        """Generates a list of blend catalogs of length batch_size. Each blend
+        catalog has entries numbered between 1 and max_number, corresponding
+        to overlapping objects in the blend.
+
+        Args:
+            catalog (astropy.Table.table): CatSim-like catalog.
+            sampling_function (btk.sampling_functions.SamplingFunction): An object that
+                                                                         return samples
+                                                                         from the catalog.
+            batch_size (int): Size of batches returned.
+            verbose: Whether to print additional information.
+        """
+        self.catalog = catalog
+        self.batch_size = batch_size
+        self.verbose = verbose
+        self.sampling_function = sampling_function
+
+        if not hasattr(sampling_function, "max_number"):
+            raise AttributeError("Sampling function must have attribute 'max_number'.")
+        self.max_number = self.sampling_function.max_number
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
         try:
             blend_catalogs = []
-            for i in range(Args.batch_size):
-                if sampling_function:
-                    blend_catalog = sampling_function(Args, catalog)
-                else:
-                    blend_catalog = default_sampling(Args, catalog)
-                    if Args.verbose:
-                        print(
-                            "Blends sampled from the catalog with the \
-                              default random sampling function"
-                        )
-                if len(blend_catalog) > Args.max_number:
+            for i in range(self.batch_size):
+                blend_catalog = self.sampling_function(self.catalog)
+                if len(blend_catalog) > self.max_number:
                     raise ValueError(
                         "Number of objects per blend must be "
                         "less than max_number: {0} <= {1}".format(
-                            len(blend_catalog), Args.max_number
+                            len(blend_catalog), self.max_number
```

Comment:
If I understand correctly, the limit from `max_number` should be taken care of in a `sampling_function()` defined by a user. In the exception message, it might be nice to suggest a user should fix the `sampling_function()`.

Reply:
Done, thanks!